src/EditorFeatures/Core/Tagging/AbstractAsynchronousTaggerProvider.TagSource_ProduceTags.cs

```diff
@@ -262,29 +269,24 @@ private void EnqueueWork(bool highPriority, bool frozenPartialSemantics, Cancell
                 // bails gracefully by checking below.
                 await _visibilityTracker.DelayWhileNonVisibleAsync(
                     _dataSource.ThreadingContext, _dataSource.AsyncListener, _subjectBuffer, DelayTimeSpan.NonFocus, cancellationToken).NoThrowAwaitable(captureContext: true);
-
-                if (cancellationToken.IsCancellationRequested)
-                    return default;
             }
 
-            _dataSource.ThreadingContext.ThrowIfNotOnUIThread();
-            if (cancellationToken.IsCancellationRequested)
-                return default;
-
             using (Logger.LogBlock(FunctionId.Tagger_TagSource_RecomputeTags, cancellationToken))
             {
+                _dataSource.ThreadingContext.ThrowIfNotOnUIThread();
+                if (cancellationToken.IsCancellationRequested)
+                    return default;
+
                 // Make a copy of all the data we need while we're on the foreground.  Then switch to a threadpool
                 // thread to do the computation. Finally, once new tags have been computed, then we update our state
                 // again on the foreground.
                 var spansToTag = GetSpansAndDocumentsToTag();
                 var caretPosition = _dataSource.GetCaretPoint(_textView, _subjectBuffer);
```

Comment:
we still grab this on the foreground.

Reply:
I wonder how hard it would be to also allow the editor to make this free-threaded.

Reply:
we could have a separate convo about that with the editor in teh future.  :)   it came up in teh perf meeting today.